src/os.c

```diff
@@ -204,6 +213,19 @@ static void* mi_win_virtual_alloc(void* addr, size_t size, size_t try_alignment,
   return p;
 }
 
+#elif defined(__wasi__)
+static void* mi_wasm_heap_grow(size_t size, size_t try_alignment) {
+  uintptr_t base = __builtin_wasm_memory_size(0) * os_page_size;
+  uintptr_t aligned_base = _mi_align_up(base, (uintptr_t) try_alignment);
+  size_t alloc_size = aligned_base - base + size;
+  mi_assert(alloc_size >= size);
+  if (alloc_size < size) return NULL;
+  if (__builtin_wasm_memory_grow(0, alloc_size / os_page_size) == SIZE_MAX) {
+    errno = ENOMEM;
+    return NULL;
+  }
+  return (void*) aligned_base;
```

Comment:
There is a subtle race condition here: between the `__builtin_wasm_memory_size` call and the `__builtin_wasm_memory_grow` call, other threads could do `__builtin_wasm_memory_grow`. We should return something based on the return value of `__builtin_wasm_memory_grow`.